### DISCUSSION TITLE
Modified the files in node_modules

### DISCUSSION THREAD

**Discussion by ttionya** [pnpm/pnpm#3433] [Q&A] [answered]
Many thanks to the `pnpm` project for bringing me a lot of convenience.

I have a lot of projects that look similar and contain a lot of dependencies, each with about 1.5G+ of dependencies. So `pnpm` helps me to save a lot of disk space.

Now I have a problem, sometimes I edit files in `node_modules` for **easier debugging**, like adding `debugger` or `console.log` to `webpack` when developing webpack plugins (this may not be a good habit, but it does come in handy in many cases).

At this point, because of the hard link, all webpacks of projects installed with `pnpm` are modified, and the files in the `.pnpm-store` are modified as well. When I finish debugging, sometimes I may not restore the changes to the files in the `node_modules` and I will just run `pnpm install` to update the latest dependencies. This way the files in the `.pnpm-store` are restored, but the `inode` is changed, while the other 20+ projects still use the old `inode` for webpack, so I need to run `pnpm install` once in all projects to eliminate the possible problem.

`webpack` is an example, sometimes for simple debugging I may modify files in any dependency.

I wonder if anyone has encountered a similar problem, thanks for joining the discussion.

**Comment by zkochan** [comment, marked as answer]
Yes, it is how it currently works. There is no solution to this now unless your system supports copy-on-write or you choose copying instead of hard linking.

**Reply by ttionya** [reply]
Hey @zkochan ,

Thank you for your reply.

I checked that MacOS APFS file system should support "copy on write", but it doesn't work and I checked #1505, #2761, libuv/libuv#2936 and libuv/libuv#2578. it might be fixed later. But the windows NTFS file system, which has a lot of developers, does not support "copy on write".

As pnpm says, "Fast, disk space efficient". I believe many developers use pnpm because of this. And using coping instead of hard link or clone would break the definition of pnpm.

Now pnpm will recreate the files in the `.pnpm-store`, causing the inode to change and the hard link to fail. **Is it possible to overwrite the contents of the `.pnpm-store` files so that inode changes can be avoided and all repositories do not need to be re-run for `pnpm install`?** Or will this treatment cause other problems?

Looking forward to your answer, thank you.

**Reply by zkochan** [reply]
hm, interesting idea. It might work. Could you create an issue for this?

**Reply by ttionya** [reply]
@zkochan ,

Thank you for your interest in this, I have created an issue #3445 .
